### PR TITLE
WalletFactory survive upgrades and repair outstanding purses and offers

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
closes: #8488
closes: #8292
closes: #8293
closes: #8286
refs: #8445
refs: #8609 
refs: #8445
refs: #8557
refs: [a3p#34](https://github.com/Agoric/agoric-3-proposals/pull/34)

## Description

Fix the WalletFactory to survive upgrade, and repair all outstanding offers and purses.

### Security Considerations

Ensure survival of outstanding SmartWallets.

### Scaling Considerations

Some work is done for existing purses and offers when WalletFactory is upgraded to incarnation 2. It's a small amount of work for each, and there aren't many on the chain at this point.

### Documentation Considerations

No changes to user-visible behavior are intended.

### Testing Considerations

Some unit tests are included. Some tests that the upgrade succeeds in A3P are here. Further tests from #8557 will be added separately. Some manual tests in the A3P environment will be automated and also added separately.

### Upgrade Considerations

Oh, yeah. It's all about upgrading.

### CODE REVIEW

File: packages/builders/scripts/smart-wallet/build-wallet-factory2-upgrade.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { makeHelpers } from '@agoric/deploy-script-support';
```
Comment (dckc) [inline packages/builders/scripts/smart-wallet/build-wallet-factory2-upgrade.js]: I don't see a `builders` package on the branch that this is going into.
I think this needs to be somewhere with a `package.json` to work.

File: packages/inter-protocol/test/smartWallet/test-oracle-integration.js [hunk 2]
```diff
@@ -149,7 +149,7 @@ const acceptInvitation = async (wallet, priceAggregator) => {
 
 let pushPriceCounter = 0;
 /**
- * @param {*} wallet
+ * @param {import('@agoric/smart-wallet/src/smartWallet.js').SmartWallet} wallet
```
Comment (dckc) [inline packages/inter-protocol/test/smartWallet/test-oracle-integration.js:152]: MH usually avoids changes that have no runtime impact on the release branch. I'm not sure whether that matters here.

File: packages/smart-wallet/package.json [hunk 3]
```diff
@@ -26,6 +27,7 @@
   "dependencies": {
     "@agoric/assert": "^0.6.1-u11wf.0",
     "@agoric/casting": "^0.4.3-u13.0",
+    "@agoric/deploy-script-support": "^0.10.4-u13.0",
```
Comment (dckc) [inline packages/smart-wallet/package.json:30]: I wonder if that could be moved to `devDependencies`.
not sure it matters, much, though...

File: packages/smart-wallet/src/proposals/upgrade-wallet-factory2-proposal.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ * @param {object} options
+ * @param {{ walletRef: VatSourceRef }} options.options
+ */
+export const upgradeWalletFactory = async (
+  {
+    consume: {
+      walletFactoryStartResult,
+      provisionPoolStartResult,
+      chainStorage,
+      walletBridgeManager: walletBridgeManagerP,
```
Comment (dckc) [inline packages/smart-wallet/src/proposals/upgrade-wallet-factory2-proposal.js:16]: :clap: This is a pretty clean set of bootstrap powers to be using.

**Review by dckc — COMMENTED** [review]
more after a short meeting...

**Review by dckc — APPROVED** [review]
looks like there's an extra copy of a file. not fatal

File: packages/builders/scripts/smart-wallet/build-wallet-factory2-upgrade.js (hunk 1 above)
Comment (Chris-Hibbert) [reply packages/builders/scripts/smart-wallet/build-wallet-factory2-upgrade.js]: It's a duplicate of one in vats anyway.  Gone.

File: packages/inter-protocol/test/smartWallet/test-oracle-integration.js (hunk 2 above)
Comment (mhofman) [reply packages/inter-protocol/test/smartWallet/test-oracle-integration.js:152]: not a problem in test files

File: packages/smart-wallet/package.json (hunk 3 above)
Comment (mhofman) [reply packages/smart-wallet/package.json:30]: Good question, which we can punt on for now.

**Review by mhofman — COMMENTED** [review]
> Some tests that the upgrade succeeds in A3P are here. Further tests from https://github.com/Agoric/agoric-sdk/pull/8557 will be added separately. Some manual tests in the A3P environment will be automated and also added separately.

I am confused. I do not see any a3p-integration changes here, so what exactly is related to A3P? I understand the wallet factory work may not be complete, but from what I understand, the upgrade of the wallet factory vat with the core proposal included in this PR should go through, right? Can we at least add that to a3p-integration in this PR, something along the lines of https://github.com/Agoric/agoric-sdk/commit/65d8c5b86b5e4a076185dca5297145cd113b4f4c

**Comment by mhofman** [comment]
> Can we at least add that to a3p-integration in this PR

I have done just that in [016da83](https://github.com/Agoric/agoric-sdk/pull/8773/commits/016da83757ff316de197eb39e91955cab4f0159d), and a crude test seems to pass!

https://github.com/Agoric/agoric-sdk/actions/runs/7578792312/job/20641968845?pr=8773#step:6:1012

https://github.com/Agoric/agoric-sdk/actions/runs/7578792312/job/20641968845?pr=8773#step:7:196

**Comment by Chris-Hibbert** [comment]
>> Some tests that the upgrade succeeds in A3P are here. Further tests from https://github.com/Agoric/agoric-sdk/pull/8557 will be added separately. Some manual tests in the A3P environment will be automated and also added separately.

> I am confused. I do not see any a3p-integration changes here, so what exactly is related to A3P? I understand the wallet factory work may not be complete, but from what I understand, the upgrade of the wallet factory vat with the core proposal included in this PR should go through, right? Can we at least add that to a3p-integration in this PR, something along the lines of https://github.com/Agoric/agoric-sdk/commit/65d8c5b86b5e4a076185dca5297145cd113b4f4c

I ran manuall tests in a3p-integration based on a proposals I wrote myself. I didn't have a version of a3p-integration that added upgrade-14, so it didn't make sense to check them in. Since you have that, it's a good addition. Your simple test is the same thing I used as a first cut. I'm working on converting the manual tests to ones I can script (shell vs. js invocations). My expectation was that they would go in master, but I can bring them here as well once there's a here to target.

**Comment by mhofman** [comment]
> I ran manuall tests in a3p-integration based on a proposals I wrote myself. I didn't have a version of a3p-integration that added upgrade-14, so it didn't make sense to check them in. Since you have that, it's a good addition. Your simple test is the same thing I used as a first cut. I'm working on converting the manual tests to ones I can script (shell vs. js invocations). My expectation was that they would go in master, but I can bring them here as well once there's a here to target.

Sorry for the miscommunication, the whole point of the `tmp-upgrade-14` was to provide that working a3p-integration environment, supporting core-proposal during upgrade. The same should work in `master` since it's based off master code I had to cherry-pick. Similarly, I would expect most of this PR to be an adaptation of code in master or about to be in master.

**Comment by Chris-Hibbert** [comment]
> I would expect most of this PR to be an adaptation of code in master or about to be in master.

#8775 will merge all the new code back to master.

**Pull request merged**